crates/compiler/src/compiler.rs

```diff
@@ -101,14 +109,35 @@ fn find_tracking_nodes(mut state: CompilationIntermediate) -> CompilationInterme
     state
 }
 
+fn check_types(mut state: CompilationIntermediate) -> CompilationIntermediate {
+    for file in &state.parsed_files {
+        let mut visitor = TypeCheckVisitor::new(
+            file.name.clone(),
+            state.known_variable_declarations.clone(),
+            file.tokens(),
+        );
+        visitor.visit(&*file.tree);
```

Comment:
Why the `&*`?
Custom deref implementation?

Reply:
antlr4rust does something weird with a trait called `DerefSeal`. Their docs say that this is the right way and I didn't question it too much. But yeah, looks like they're doing some deref magic under the hood to make sure the right visit step is called.

Reply:
Yea, I have seen similar stuff for Bevy when deref is used. I don't like how it looks, but sometimes it's worth it for the ergonomic wins for other uses.
Anyways, let's leave it as-is for now :)